exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java

```diff
@@ -85,109 +91,231 @@ protected ConvertCountToDirectScan(RelOptRuleOperand rule, String id) {
+      RelDataTypeField field = new RelDataTypeFieldImpl("count$" + i, i, aggregateRel.getCluster().getTypeFactory().createSqlType(SqlTypeName.BIGINT));
+      fields.add(field);
     }
-
+    return new RelRecordType(fields);
   }
 
   /**
-   * Class to represent the count aggregate result.
+   * Builds schema based on given field names.
+   * Type for each schema is set to long.class.
+   *
+   * @param fieldNames field names
+   * @return schema
    */
-  public static class CountQueryResult {
-    public long count;
+  private LinkedHashMap<String, Class<?>> buildSchema(List<String> fieldNames) {
+    LinkedHashMap<String, Class<?>> schema = new LinkedHashMap<>();
+    for (String fieldName: fieldNames) {
+      schema.put(fieldName, long.class);
+    }
+    return schema;
+  }
 
-    public CountQueryResult(long cnt) {
-      this.count = cnt;
+  /**
+   * For each field creates row expression.
+   *
+   * @param rowType row type
+   * @return list of row expressions
+   */
+  private List<RexNode> prepareFieldExpressions(RelDataType rowType) {
+    List<RexNode> expressions = new ArrayList<>();
+    for (int i = 0; i < rowType.getFieldCount(); i++) {
+      expressions.add(RexInputRef.of(i, rowType));
     }
+    return expressions;
   }
 
-  private RelDataType getCountDirectScanRowType(RelDataTypeFactory typeFactory) {
-    List<RelDataTypeField> fields = Lists.newArrayList();
-    fields.add(new RelDataTypeFieldImpl("count", 0, typeFactory.createSqlType(SqlTypeName.BIGINT)));
+  /**
+   * Helper class to collect counts based on metadata information.
+   * For example, for parquet files it can be obtained from parquet footer (total row count)
+   * or from parquet metadata files (column counts).
+   */
+  private class CountsCollector {
+
+    private final PlannerSettings settings;
+    private final Set<String> implicitColumnsNames;
+    private final List<SchemaPath> columns;
+    private final List<Long> counts;
+
+    CountsCollector(PlannerSettings settings) {
+      this.settings = settings;
+      this.implicitColumnsNames = ColumnExplorer.initImplicitFileColumns(settings.getOptions()).keySet();
+      this.counts = new ArrayList<>();
+      this.columns = new ArrayList<>();
+    }
 
-    return new RelRecordType(fields);
-  }
+    /**
+     * Collects counts for each aggregation call.
+     * Will fail to collect counts if:
+     * <ol>
+     *   <li>was not able to determine count for at least one aggregation call</li>
+     *   <li>count if used for file system partition column</li>
+     * </ol>
+     *
+     * @param agg aggregate relational expression
+     * @param scan scan relational expression
+     * @param project project relational expression
+     *
+     * @return true if counts were collected, false otherwise
+     */
+    boolean collect(DrillAggregateRel agg, DrillScanRel scan, DrillProjectRel project) {
+      return calculateCounts(agg, scan, project) && !containsPartitionColumns();
```

Comment:
If user calls collect(agg, scan, project), the class should clear/reset all the internal collections like "columns", "counts".  Otherwise, what if user uses the same CountsCollector instance, and call collect() twice with different inputs?